cpp/src/lists/copying/concatenate.cu

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <algorithm>
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/concatenate.hpp>
+#include <cudf/detail/concatenate.cuh>
+#include <cudf/lists/lists_column_view.hpp>
+#include <memory>
+
+//#include <../tests/utilities/column_utilities.hpp>
+
+namespace cudf {
+namespace lists {
+namespace detail {
+
+namespace {
+
+/**
+ * @brief Merges the offsets child columns of multiple list columns into one.
+ *
+ * Since offsets are all relative to the start of their respective column,
+ * all offsets are shifted to account for the new starting position
+ *
+ * @param columns               Vector of lists columns to concatenate
+ * @param total_list_count      Total number of lists contains in the columns
+ * @param print_all_differences If true display all differences
+ **/
+std::unique_ptr<column> merge_offsets(std::vector<lists_column_view> const& columns,
```

Comment:
Right, so I looked at that.  The strings concatenate switched over to using the "fused kernel" method where the chars, offsets and validity are copied in a single kernel launch (https://github.com/rapidsai/cudf/blob/66a77e35f517d9b52f26a688ee7ef7b9d4693c66/cpp/src/strings/copying/concatenate.cu#L123).  

The structure of lists doesn't really fuse that way (since the "data" can simply be further nested lists).  However, it would be a reasonable optimization to combine the offset and validity concatenation for a given level into a single kernel.   Should I open an issue for that?

Reply:
Sure. However I didn't notice that, I just assumed the offset merging could be handled the same -- but I wasn't thinking about multilevel nesting.